edx_django_utils/monitoring/middleware.py

```diff
@@ -28,53 +24,71 @@
 
 
 _DEFAULT_NAMESPACE = 'edx_django_utils.monitoring'
-_REQUEST_CACHE_NAMESPACE = '{}.custom_metrics'.format(_DEFAULT_NAMESPACE)
+_REQUEST_CACHE_NAMESPACE = '{}.custom_attributes'.format(_DEFAULT_NAMESPACE)
 
 
-class MonitoringCustomMetricsMiddleware(MiddlewareMixin):
+class CachedCustomMonitoringMiddleware(MiddlewareMixin):
     """
-    The middleware class for adding custom metrics.
+    The middleware class for handling cached custom attributes, batch reported
+    at the end of a request.
 
     Make sure to add below the request cache in MIDDLEWARE.
+
+    This middleware will only call on the newrelic agent if there are any attributes
+    to report for this request, so it will not incur any processing overhead for
+    request handlers which do not record custom attributes.
+
+    Note: New Relic adds custom attributes to events, which is what is being used here.
+
     """
+    REQUIRED_MIDDLEWARE = [
+        'edx_django_utils.cache.middleware.RequestCacheMiddleware',
+        'edx_django_utils.monitoring.middleware.CachedCustomMonitoringMiddleware',
+    ]
 
     def __init__(self, *args, **kwargs):
-        super(MonitoringCustomMetricsMiddleware, self).__init__(*args, **kwargs)
+        super(CachedCustomMonitoringMiddleware, self).__init__(*args, **kwargs)
         # checks proper dependency order as well.
-        _check_middleware_dependencies(self, required_middleware=[
-            'edx_django_utils.cache.middleware.RequestCacheMiddleware',
-            'edx_django_utils.monitoring.middleware.MonitoringCustomMetricsMiddleware',
-        ])
+        _check_middleware_dependencies(self, required_middleware=self.REQUIRED_MIDDLEWARE)
 
     @classmethod
-    def _get_metrics_cache(cls):
+    def _get_attributes_cache(cls):
         """
         Get a reference to the part of the request cache wherein we store New
-        Relic custom metrics related to the current request.
+        Relic custom attributes related to the current request.
         """
         return RequestCache(namespace=_REQUEST_CACHE_NAMESPACE)
 
     @classmethod
-    def accumulate_metric(cls, name, value):
+    def accumulate_attribute(cls, name, value):
         """
-        Accumulate a custom metric (name and value) in the metrics cache.
+        Accumulate a custom attribute (name and value) in the attributes cache.
         """
-        metrics_cache = cls._get_metrics_cache()
-        metrics_cache.setdefault(name, 0)
-        metrics_cache.set(name, value)
+        attributes_cache = cls._get_attributes_cache()
+        attributes_cache.setdefault(name, 0)
+        attributes_cache.set(name, value)
+
+    @classmethod
+    def accumulate_metric(cls, name, value):  # pragma: no cover
+        """
+        Deprecated method replaced by accumulate_attribute.
+        """
+        msg = "Use 'accumulate_attribute' in place of 'accumulate_metric'."
+        warnings.warn(msg, DeprecationWarning)
+        cls.accumulate_attribute(name, value)
```

Comment:
@timmc-edx: I'm wondering about whether this is wasting time.  This method was never meant to be public (used outside the app), and a quick search shows that it doesn't seem to be used anywhere else: https://github.com/search?q=org%3Aedx+accumulate_metric&type=Code 

I am leaning towards removing this deprecated method now, and not merging it.  I can add a note about this in the CHANGELOG, but I don't want to use a major version change, because according to the API (but, not the rules of Python), this is a backward compatible change.

Thoughts?

Reply:
I would just merge it and then remove it later in the next major version bump. Doesn't seem harmful to leave in.

Reply:
Will do.  I've got a bug fix landing in another PR and will rebase and land this afterwards.  Thanks.